File: test/integration/Assets.fs
module Tests.Integration.Assets

open System.Net.Http

open Swensen.Unquote
open Xunit


open Oryx
open CogniteSdk
open CogniteSdk.Assets
open Common
open Tests

[<Fact>]
let ``List assets with limit is Ok`` () = async {
    // Arrange
    let ctx = readCtx ()
    let query = [ AssetQuery.Limit 10 ]

    // Act
    let! res = Items.listAsync query [] ctx

    let len =
        match res.Result with
        | Ok dtos -> Seq.length dtos.Items
        | Error _ -> 0

    // Assert
    test <@ Result.isOk res.Result @>
    test <@ len = 10 @>
    test <@ res.Request.Method = HttpMethod.Post @>
    test <@ res.Request.Extra.["resource"] = "/assets/list" @>
}

[<Fact>]
let ``Get asset by id is Ok`` () = async {
    // Arrange
    let ctx = readCtx ()
    let assetId = 130452390632424L

    // Act
    let! res = Assets.Entity.getAsync assetId ctx

    let resId =
        match res.Result with
        | Ok dto -> dto.Id
        | Error _ -> 0L

    // Assert
    test <@ Result.isOk res.Result @>
    test <@ resId = assetId @>
    test <@ res.Request.Method = HttpMethod.Get @>
    test <@ res.Request.Extra.["resource"] = "/assets/130452390632424" @>
}

[<Fact>]
let ``Get asset by missing id is Error`` () = async {
    // Arrange
    let ctx = readCtx ()
    let assetId = 0L

    // Act
    let! res = Assets.Entity.getAsync assetId ctx

    let err =
        match res.Result with
        | Ok _ -> ResponseError.empty
        | Error err -> err

    // Assert
    test <@ Result.isError res.Result @>
    test <@ err.Code = 400 @>
    test <@ err.Message = "getAsset.arg0: must be greater than or equal to 1" @>
}

[<Fact>]
let ``Get asset by ids is Ok`` () = async {
    // Arrange
    let ctx = readCtx ()
    let assetIds =
        [ 130452390632424L; 126847700303897L; 124419735577853L ]
        |> Seq.map Identity.Id

    // Act
    let! res = Assets.Retrieve.getByIdsAsync assetIds ctx

    let len =
        match res.Result with
        | Ok dtos -> Seq.length dtos
        | Error _ -> 0

    // Assert
    test <@ Result.isOk res.Result @>
    test <@ len = 3 @>
    test <@ res.Request.Method = HttpMethod.Post @>
    test <@ res.Request.Extra.["resource"] = "/assets/byids" @>
}

[<Fact>]
let ``Search assets is Ok`` () = async {
    // Arrange
    let ctx = readCtx ()
    let options = [
        AssetSearch.Name "23"
    ]

    // Act
    let! res = Assets.Search.searchAsync 10 options [] ctx

    let len =
        match res.Result with
        | Ok dtos -> Seq.length dtos
        | Error _ -> 0

    // Assert
    test <@ Result.isOk res.Result @>
    test <@ len = 10 @>
    test <@ res.Request.Method = HttpMethod.Post @>
    test <@ res.Request.Extra.["resource"] = "/assets/search" @>
}

[<Fact>]
let ``Filter assets is Ok`` () = async {
    // Arrange
    let ctx = readCtx ()
    let options = [
        AssetQuery.Limit 10
    ]
    let filters = [
        AssetFilter.RootIds [ Identity.Id 6687602007296940L ]
    ]

    // Act
    let! res = Assets.Items.listAsync options filters ctx

    let len =
        match res.Result with
        | Ok dtos -> Seq.length dtos.Items
        | Error _ -> 0

    // Assert
    test <@ Result.isOk res.Result @>
    test <@ len = 10 @>
    test <@ res.Request.Method = HttpMethod.Post @>
    test <@ res.Request.Extra.["resource"] = "/assets/list" @>
}

[<Fact>]
let ``Create and delete assets is Ok`` () = async {
    // Arrange
    let ctx = writeCtx ()
    let externalIdString = "createDeleteTestAssets"
    let dto: AssetWriteDto = {
        ExternalId = Some externalIdString
        Name = "Create Assets sdk test"
        ParentId = None
        Description = Some "dotnet sdk test"
        MetaData = Map.empty
        Source = None
        ParentExternalId = None
    }
    let externalId = Identity.ExternalId externalIdString

    // Act
    let! res = Assets.Create.createAsync [ dto ] ctx
    let! delRes = Assets.Delete.deleteAsync ([ externalId ], false) ctx
    let resExternalId =
        match res.Result with
        | Ok assetsResponses ->
            let h = Seq.tryHead assetsResponses
            match h with
            | Some assetsResponse -> assetsResponse.ExternalId
            | None -> None
        | Error _ -> None

    // Assert
    test <@ Result.isOk res.Result @>
    test <@ resExternalId = Some externalIdString @>
    test <@ res.Request.Method = HttpMethod.Post @>
    test <@ res.Request.Extra.["resource"] = "/assets" @>
    test <@ res.Request.Query.IsEmpty @>

    test <@ Result.isOk delRes.Result @>
    test <@ delRes.Request.Method = HttpMethod.Post @>
    test <@ delRes.Request.Extra.["resource"] = "/assets/delete" @>
    test <@ delRes.Request.Query.IsEmpty @>
}

[<Fact>]
let ``Update assets is Ok`` () = async {
    // Arrange
    let wctx = writeCtx ()

    let externalIdString = "createDeleteTestAssets"
    let newMetadata = ([
        "key1", "value1"
        "key2", "value2"
    ]
    |> Map.ofList)
    let dto: AssetWriteDto = {
        ExternalId = Some externalIdString
        Name = "Create Assets sdk test"
        ParentId = None
        Description = Some "dotnet sdk test"
        MetaData = [
            "oldkey1", "oldvalue1"
            "oldkey2", "oldvalue2"
        ] |> Map.ofList
        Source = None
        ParentExternalId = None
    }
    let externalId = Identity.ExternalId externalIdString
    let newName = "UpdatedName"
    // Act
    let! createRes = Assets.Create.createAsync [ dto ] wctx
    let! updateRes =
        Assets.Update.updateAsync [
            (externalId, [
                AssetUpdate.SetName newName
                AssetUpdate.ChangeMetaData (newMetadata, [ "oldkey1" ] |> Seq.ofList)
            ])
        ] wctx
    let! getRes = Assets.Retrieve.getByIdsAsync [ externalId ] wctx
    let! delRes = Assets.Delete.deleteAsync ([ externalId ], false) wctx

    let resName, resExternalId, resMetaData =
        match getRes.Result with
        | Ok assetsResponses ->
            let h = Seq.tryHead assetsResponses
            match h with
            | Some assetResponse -> assetResponse.Name, assetResponse.ExternalId, assetResponse.MetaData
            | None -> "", Some "", Map.empty
        | Error _ -> "", Some "", Map.empty

    let updateSuccsess =
        match updateRes.Result with
        | Ok res -> true
        | Error _ -> false

    let metaDataOk =
        resMetaData.ContainsKey "key1"
        && resMetaData.ContainsKey "key2"
        && resMetaData.ContainsKey "oldkey2"
        && not (resMetaData.ContainsKey "oldkey1")

    // Assert create
    test <@ Result.isOk createRes.Result @>
    test <@ createRes.Request.Method = HttpMethod.Post @>
    test <@ createRes.Request.Extra.["resource"] = "/assets" @>
    test <@ createRes.Request.Query.IsEmpty @>

    // Assert update
    test <@ updateSuccsess @>
    test <@ Result.isOk updateRes.Result @>
    test <@ updateRes.Request.Method = HttpMethod.Post @>
    test <@ updateRes.Request.Extra.["resource"] = "/assets/update" @>
    test <@ updateRes.Request.Query.IsEmpty @>

    // Assert get
    test <@ Result.isOk getRes.Result @>
    test <@ getRes.Request.Method = HttpMethod.Post @>
    test <@ getRes.Request.Extra.["resource"] = "/assets/byids" @>
    test <@ getRes.Request.Query.IsEmpty @>
    test <@ resExternalId = Some externalIdString @>
    test <@ resName = newName @>
    test <@ metaDataOk @>

    // Assert delete
    test <@ Result.isOk delRes.Result @>
    test <@ delRes.Request.Method = HttpMethod.Post @>
    test <@ delRes.Request.Extra.["resource"] = "/assets/delete" @>
    test <@ delRes.Request.Query.IsEmpty @>
}

File: test/integration/csharp/Assets.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Collections.Generic;
using System.Text;
using System.Linq;
using CogniteSdk.Assets;
using System.Threading.Tasks;
using CogniteSdk;
using System.Diagnostics;

namespace Test.CSharp.Integration {

    [TestClass]
    public class Assets : TestBase {

        [TestMethod]
        [TestCategory(Category.Asset)]
        [Description("Ensures listing Assets returns number of assets equal to 'limit' value")]

        public void ListingAssetsRespectsLimit() {
            // Arrange
            var limit = 10;
            var option = AssetQuery.Limit(limit);

            // Act
            var res = ReadClient.Assets.ListAsync(new List<AssetQuery> { option });

            // Assert
            Assert.AreEqual(limit, res.Result.Items.Count(), "Expected the number of assets to be the same as the Limit value");
        }

        [TestMethod]
        [TestCategory(Category.Asset)]
        [Description("Ensures that getting an asset by ID returns the correct asset")]

        public void AssetByIdReturnsCorrectAsset() {
            // Arrange
            var assetId = 130452390632424;

            // Act
            var res = ReadClient.Assets.GetAsync(assetId).Result;

            // Assert
            Assert.AreEqual(assetId, res.Id, "The received Asset doesn't match the ID");
        }

        [TestMethod]
        [TestCategory(Category.Asset)]
        [Description("Ensures that getting an asset with an invalid Id doesnt return anything")]

        public async Task AssetByInvalidIdReturnsErrorAsync() {
            // Arrange
            var assetId = 0;
            bool caughtException = false;

            // Act
            try {
                var res = await ReadClient.Assets.GetAsync(assetId);
            } catch (ResponseException) {
                caughtException = true;
            }

            // Assert
            Assert.IsTrue(caughtException, "Expected request to fail with 'CogniteSdk.ResponseException' but didnt");
        }

        [TestMethod]
        [TestCategory(Category.Asset)]
        [Description("Gets multiple Assets with a list of ids and ensures theyre the correct assets")]
        public void AssetsByIdsRetrivesTheCorrectAssets() {
            // Arrange
            var ids = new List<long>() {
                130452390632424,
                126847700303897,
                124419735577853
            };

            // Act
            var res = ReadClient.Assets.GetByIdsAsync(ids).Result;

            // Assert
            var returnedIds = res.Select(asset => asset.Id);
            var resCount = res.Count();
            Assert.AreEqual(ids.Count, resCount, $"Expected {ids.Count} assets but got {resCount}");
            Assert.AreEqual(returnedIds.Intersect(ids).Count(), returnedIds.Count(), "One of the received Assets dont match the requested IDs");
        }

        [TestMethod]
        [TestCategory(Category.Asset)]
        [Description("Search Asset returns the correct number of assets")]
        public void AssetSearchReturnsExpectedNumberOfAssets() {
            // Arrange
            var options = new List<AssetSearch>() {
                AssetSearch.Name("23")
            };
            var numOfAssets = 10;

            // Act
            var res = ReadClient.Assets.SearchAsync(numOfAssets, options).Result;

            // Assert
            var resCount = res.Count();
            Assert.AreEqual(numOfAssets, resCount, $"Expected {options.Count} assets but got {resCount}");
        }

        [TestMethod]
        [TestCategory(Category.Asset)]
        [Description("Listing Assets with filter returns the expected number of assets")]
        public void FilterAssetsReturnsTheExpectedNumberOfAssets() {
            // Arrange
            var numOfAssets = 10;
            var id = 6687602007296940;
            var options = new List<AssetQuery>() {
                AssetQuery.Limit(numOfAssets)
            };
            var filter = new List<AssetFilter>() {
                AssetFilter.RootIds(new List<Identity>() { Identity.Id(id) })
            };

            // Act
            var res = ReadClient.Assets.ListAsync(options, filter).Result;

            // Assert
            var resCount = res.Items.Count();
            Assert.AreEqual(numOfAssets, resCount, $"Expected {options.Count} assets but got {resCount}");
        }

        [TestMethod]
        [TestCategory(Category.Asset)]
        [Description("Creating an asset and deletes it works")]
        public void CreateAndDeleteAssetWorkAsExpected() {
            // Arrange
            var externalIdString = "createDeleteTestAssets";
            var newAsset = new AssetEntity();
            newAsset.ExternalId = externalIdString;
            newAsset.Name = "Create Assets c# sdk test";
            newAsset.Description = "Just a test";


            // Act
            var res = WriteClient.Assets.CreateAsync(new List<AssetEntity>() { newAsset }).Result;
            WriteClient.Assets.DeleteAsync(new List<string>() { externalIdString }, false);

            // Assert
            var resCount = res.Count();
            Assert.AreEqual(1, resCount, $"Expected 1 created asset but got {resCount}");
            Assert.AreEqual(externalIdString, res.First().ExternalId, "Created externalId doesnt match expected");
        }

        [TestMethod]
        [TestCategory(Category.Asset)]
        [Description("Deleting an asset that exist fails with ResponseException")]
        public async Task AssetDeleteFailsWhenIdIsInvalidAsync() {
            // Arrange
            var id = 0;
            var caughtException = false;

            // Act
            try {
                await WriteClient.Assets.DeleteAsync(new List<long>() { id }, false);
            } catch (ResponseException) {
                caughtException = true;
            }

            // Assert
            Assert.IsTrue(caughtException, "Expected request to fail with 'CogniteSdk.ResponseException' but didnt");
        }

        [TestMethod]
        [TestCategory(Category.Asset)]
        [Description("Deleting an asset that exist fails with ResponseException")]
        public async Task UpdatedAssetsPerformsExpectedChangesAsync() {
            // Arrange
            var externalIdString = "updateAssetTest";
            var newAsset = new AssetEntity();
            var newMetadata = new Dictionary<string, string>() {
                { "key1", "value1" },
                { "key2", "value2" }
            };
            newAsset.ExternalId = externalIdString;
            newAsset.Name = "Update Assets c# sdk test";
            newAsset.Description = "Just a test";
            newAsset.MetaData = new Dictionary<string, string>() {
                { "oldkey1", "oldvalue1" },
                { "oldkey2", "oldvalue2" }
            };

            var newName = "Updated update asset";

            // Act
            var res = await WriteClient.Assets.CreateAsync(new List<AssetEntity>() { newAsset });
            await WriteClient.Assets.UpdateAsync(new List<(Identity, IEnumerable<AssetUpdate>)>() {
                (Identity.ExternalId(externalIdString), new List<AssetUpdate>() {
                    AssetUpdate.SetName(newName),
                    AssetUpdate.ChangeMetaData(newMetadata, new List<string>() { "oldkey1" })
                })
            });

            var getRes = await WriteClient.Assets.GetByIdsAsync(new List<string>() { externalIdString });
            await WriteClient.Assets.DeleteAsync(new List<string>() { externalIdString }, false);

            // Assert
            var resCount = getRes.Count();
            Assert.IsTrue(resCount == 1, $"Expected a single Asset but got {resCount}");
            var resAsset = getRes.First();
            Assert.AreEqual(externalIdString, resAsset.ExternalId, $"Asset doest have expected ExternalId. Was '{resAsset.ExternalId}' but expected '{externalIdString}'");
            Assert.AreEqual(newName, resAsset.Name, $"Expected the Asset name to update to '{newName}' but was '{resAsset.Name}'");
            Assert.IsTrue(resAsset.MetaData.ContainsKey("key1") && resAsset.MetaData.ContainsKey("key2"), "Asset wasnt update with new metadata values");
            Assert.IsTrue(resAsset.MetaData.ContainsKey("oldkey2") && !resAsset.MetaData.ContainsKey("oldkey1"), "Asset update changed unintended metadata values");
        }


    }
    
}
